In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

4 8
12

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

4 8
12


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5
1 2 3 4 5
20


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [263]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [265]:
!./max_pairwise_product_full < input1.txt

90


In [264]:
%%writefile output1.txt
90

Overwriting output1.txt


In [266]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [267]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [268]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [269]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Overwriting input3.txt


In [ ]:
200000*200000

40000000000

In [270]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [271]:
!./max_pairwise_product_full < input3.txt

1345294336


In [272]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [273]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [274]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [275]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [276]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [277]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [278]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [279]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



----------------------------------------------------**Test *3.1***-------------------------------------------------

In [465]:
%%writefile input4.txt
10
7 8 9554  5214  42  567 331 44  4 1

Overwriting input4.txt


In [466]:
9554*5214

49814556

In [467]:
%%writefile ctest4.txt
49814556

Overwriting ctest4.txt


In [468]:
!./max_pairwise_product_long_full < input4.txt

49814556


In [469]:
%%writefile output4.txt
49814556

Overwriting output4.txt


In [471]:
!./max_pairwise_product_long_full < input4.txt > ctest4.txt
!grep -xvFf output4.txt ctest4.txt

----------------------------------------------------Test 3.2-------------------------------------------------

In [285]:
%%writefile input5.txt
5
23  5 4455  6871  25

Overwriting input5.txt


In [286]:
4455*6871

30610305

In [287]:
%%writefile ctest5.txt
30610305

Overwriting ctest5.txt


In [288]:
!./max_pairwise_product_long_full < input5.txt

30610305


In [289]:
%%writefile output5.txt
30610305

Overwriting output5.txt


In [472]:
!./max_pairwise_product_long_full < input5.txt > ctest5.txt
!grep -xvFf output5.txt ctest5.txt

---------------------------------------------------Test 3.3-------------------------------------------------

In [291]:
%%writefile input6.txt
20
12  3 6 7 918 4 1 2 85  3 7 23  45  68  72  23  100 58  75  64

Overwriting input6.txt


In [292]:
918*100

91800

In [293]:
%%writefile ctest6.txt
91800

Overwriting ctest6.txt


In [294]:
!./max_pairwise_product_long_full < input6.txt

91800


In [295]:
%%writefile output6.txt
91800

Overwriting output6.txt


In [ ]:
!./max_pairwise_product_long_full < input6.txt > ctest6.txt
!grep -xvFf output6.txt ctest6.txt

----------------------------------------------------Test 3.4-------------------------------------------------

In [503]:
%%writefile input7.txt
3
233356464 112745 4556658

Overwriting input7.txt


In [477]:
233356464*4556658

1063325598537312

In [505]:
%%writefile ctest7.txt
1063325598537312

Overwriting ctest7.txt


In [506]:
!./max_pairwise_product_long_full < input7.txt

1063325598537312


In [507]:
%%writefile output7.txt
1063325598537312

Overwriting output7.txt


In [508]:
!./max_pairwise_product_long_full < input7.txt > ctest7.txt
!grep -xvFf output7.txt ctest7.txt

---------------------------------------------------Test 3.5-------------------------------------------------

In [509]:
%%writefile input8.txt
4
10101 10001101 101 1 2

Overwriting input8.txt


In [501]:
10101*10001101

101021121201

In [510]:
%%writefile ctest8.txt
101021121201

Overwriting ctest8.txt


In [511]:
!./max_pairwise_product_long_full < input8.txt

101021121201


In [512]:
%%writefile output8.txt
101021121201

Overwriting output8.txt


In [513]:
!./max_pairwise_product_long_full < input8.txt > ctest8.txt
!grep -xvFf output8.txt ctest8.txt

---------------------------------------------------Test 3.6-------------------------------------------------

In [514]:
%%writefile input9.txt
7
2 3 5 7 4 9 1

Overwriting input9.txt


In [494]:
7*9

63

In [515]:
%%writefile ctest9.txt
63

Overwriting ctest9.txt


In [516]:
!./max_pairwise_product_long_full < input9.txt

63


In [517]:
%%writefile output9.txt
63

Overwriting output9.txt


In [518]:
!./max_pairwise_product_long_full < input9.txt > ctest9.txt
!grep -xvFf output9.txt ctest9.txt

---------------------------------------------------Test 3.7-------------------------------------------------

In [519]:
%%writefile input10.txt
12
2 3 5 7 4 9 1 55 44 56 78 98

Overwriting input10.txt


In [314]:
78*98

7644

In [520]:
%%writefile ctest10.txt
7644

Overwriting ctest10.txt


In [521]:
!./max_pairwise_product_long_full < input10.txt

7644


In [323]:
%%writefile output10.txt
7644

Overwriting output10.txt


In [522]:
!./max_pairwise_product_long_full < input10.txt > ctest10.txt
!grep -xvFf output10.txt ctest10.txt

-----------------all tests 4-10--------------

In [523]:
 %%writefile run_all_test2.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {4..10}
do
    echo "------------------------"
    ./max_pairwise_product_long_full < input$i.txt
    differences=$(grep -xvFf output$i.txt ctest$i.txt)
    if [ -n "$differences" ]; then
        echo "Diferencias encontradas en output$i.txt and ctest$i.txt:"
        echo "$differences"
    else
        echo "input$i.txt: OK"
    fi
done

Overwriting run_all_test2.sh


In [524]:
!chmod u+x run_all_test2.sh
!./run_all_test2.sh

------------------------
49814556
input4.txt: OK
------------------------
30610305
input5.txt: OK
------------------------
91800
input6.txt: OK
------------------------
1063325598537312
input7.txt: OK
------------------------
101021121201
input8.txt: OK
------------------------
63
input9.txt: OK
------------------------
7644
input10.txt: OK


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [244]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [243]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Overwriting max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [245]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

563



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n = 10

563



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [246]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!time ./max_pairwise_product_long_full_execints < input12.txt


563

real	0m0.003s
user	0m0.002s
sys	0m0.001s


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

306/200001000010

1.5299923499617505e-09

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [247]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [248]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.856s
user	0m0.846s
sys	0m0.001s


## Python

In [257]:
# Crear y escribir en el archivo
with open("input13.txt", "w") as file:
    file.write("11000\n")  # n
    for i in range(1, 11001):
        file.write(f"{i} ")

print("Archivo input13.txt creado y escrito correctamente.")


Archivo input13.txt creado y escrito correctamente.


In [250]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    max_product = 0

    for first in range(len(numbers)):
        for second in range(len(numbers)):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    M = max_pairwise_product(numbers)
    print(M)


Overwriting max_pairwise_product_python.py


In [251]:
# Verify
!time python max_pairwise_product_python.py < input13.txt

120989000

real	0m19.457s
user	0m19.212s
sys	0m0.023s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product< input13.txt

120989000

real	0m0.596s
user	0m0.581s
sys	0m0.005s


In [253]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


Overwriting max_pairwise_product_long.cpp


In [258]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long< input13.txt

120989000

real	0m0.403s
user	0m0.400s
sys	0m0.001s


In [ ]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    print(max_pairwise_product(numbers))


Overwriting max_pairwise_product_python.py


In [254]:
!time python max_pairwise_product_python.py < input13.txt

120989000

real	0m20.997s
user	0m20.713s
sys	0m0.019s


In [255]:
%%writefile max_pairwise_product_long.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                max_product = aux_product

    return max_product


Overwriting max_pairwise_product_long.py


In [259]:
!time python max_pairwise_product_long.py < input13.txt


real	0m0.054s
user	0m0.037s
sys	0m0.009s


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [404]:
%%writefile max_pairwise_product_long_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>


int main() {                                              // 1 instrucción para iniciar la función
    int n;                                                // 1 instrucción para declarar la variable 'n'
    std::cin >> n;                                        // 1 instrucción para leer el valor de 'n' desde la entrada estándar
    std::vector<int> valores(n);                          // 1 instrucción para crear un vector de enteros 'valores' y asignarle capacidad para 'n' elementos

    for (int i = 0; i < n; ++i) {
        std::cin >> valores[i];                           // n instrucciones para leer 'n' valores desde la entrada estándar y asignarlos a elementos del vector
    }

    std::sort(valores.begin(), valores.end());            // n * lg(n) instrucciones en promedio para ordenar el vector 'valores'
    std::cout << valores[n-1] * valores[n-2] << "\n";     // 4 instrucciones para imprimir el resultado

    return 0;                                             // 1 instrucción para finalizar la función
}



Overwriting max_pairwise_product_long_sort.cpp


Test with inputs 1 and 2



In [405]:
!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!./max_pairwise_product_long_sort < input1.txt

!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!./max_pairwise_product_long_sort < input2.txt

90
90


Test and time it with input 3

In [406]:
!g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
!time ./max_pairwise_product_long_sort < input3.txt

1345294336

real	0m0.003s
user	0m0.000s
sys	0m0.003s


20 stress tests

In [407]:
%%writefile input14.txt
10
2 4 6 8 10 12 14 16 18 20

Overwriting input14.txt


In [ ]:
%%writefile ctest14.txt
360

In [409]:
%%writefile input15.txt
5
1 2 3 4 5

Overwriting input15.txt


In [ ]:
%%writefile ctest15.txt
20

In [411]:
%%writefile input16.txt
20
6 2 58 28 4 54 468 654 45  45 52 545 2 52 52625 42 54 641 416 6544

Overwriting input16.txt


In [430]:
52625*6544

344378000

In [ ]:
%%writefile ctest16.txt
344378000

In [413]:
%%writefile input17.txt
10
0 0 0 0 0 0 0 0 0  0

Overwriting input17.txt


In [433]:
%%writefile ctest17.txt
0

Overwriting ctest17.txt


In [415]:
%%writefile input18.txt
10
0 0 2 0 4 0 0 0 0  0

Overwriting input18.txt


In [436]:
%%writefile ctest18.txt
8

Overwriting ctest18.txt


In [417]:
%%writefile input19.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input19.txt


In [438]:
%%writefile ctest19.txt
90

Overwriting ctest19.txt


In [440]:
%%writefile input20.txt
5
5 4 0 1 1

Overwriting input20.txt


In [441]:
%%writefile ctest20.txt
20

Overwriting ctest20.txt


In [421]:
%%writefile input21.txt
4
3 33 333 3333

Overwriting input21.txt


In [442]:
%%writefile ctest21.txt
1109889

Overwriting ctest21.txt


In [423]:
%%writefile input22.txt
3
3 33 333

Overwriting input22.txt


In [444]:
%%writefile ctest22.txt
10989

Overwriting ctest22.txt


In [449]:
%%writefile input23.txt
20
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19

Overwriting input23.txt


In [445]:
18*19

342

In [454]:
%%writefile ctest23.txt
342

Overwriting ctest23.txt


------si quisiera mas test, con el siguiente código se podria generar----

In [575]:
import random

# Genera 20 archivos desde input14 hasta input34
for i in range(24, 35):
    # Genera un número aleatorio de elementos en el rango de 2 a 10
    num_elements = random.randint(2, 10)

    # Genera una lista de números aleatorios en el rango de 0 a 10000
    numbers = [random.randint(0, 10000) for _ in range(num_elements)]

    # Nombre del archivo de entrada
    input_filename = f'input{i}.txt'

    # Escribe la lista de números en el archivo de entrada
    with open(input_filename, 'w') as input_file:
        input_file.write(f'{num_elements}\n')
        input_file.write(' '.join(map(str, numbers)))

    # Calcula los dos valores más grandes en la lista
    numbers.sort(reverse=True)
    largest = numbers[0]
    second_largest = numbers[1]

    # Calcula el producto de los dos valores más grandes
    product = largest * second_largest

    # Nombre del archivo de salida
    ctest_filename = f'ctest{i}.txt'

    # Escribe el producto en el archivo de salida
    with open(ctest_filename, 'w') as ctest_file:
        ctest_file.write(str(product))

    print(f'Se ha creado el archivo {input_filename}')
    print(f'Se ha creado el archivo {ctest_filename}')



Se ha creado el archivo input24.txt
Se ha creado el archivo ctest24.txt
Se ha creado el archivo input25.txt
Se ha creado el archivo ctest25.txt
Se ha creado el archivo input26.txt
Se ha creado el archivo ctest26.txt
Se ha creado el archivo input27.txt
Se ha creado el archivo ctest27.txt
Se ha creado el archivo input28.txt
Se ha creado el archivo ctest28.txt
Se ha creado el archivo input29.txt
Se ha creado el archivo ctest29.txt
Se ha creado el archivo input30.txt
Se ha creado el archivo ctest30.txt
Se ha creado el archivo input31.txt
Se ha creado el archivo ctest31.txt
Se ha creado el archivo input32.txt
Se ha creado el archivo ctest32.txt
Se ha creado el archivo input33.txt
Se ha creado el archivo ctest33.txt
Se ha creado el archivo input34.txt
Se ha creado el archivo ctest34.txt


In [547]:
8536*3576

30524736

---------------- ya creados los test se procede a realizar el stress testing ------------------------

In [633]:
 %%writefile torture_find_diference.sh
#!/bin/bash

g++ max_pairwise_product_long_sort.cpp -o max_pairwise_product_long_sort
for i in {14..34}
do
    echo "-----------------------------------"
    echo "Running find_cpp with input$i.txt"
    ./max_pairwise_product_long_sort < input$i.txt>output$i.txt
    ./max_pairwise_product_long_sort < input$i.txt
    differences=$(grep -xvFf output$i.txt ctest$i.txt)
    if [ -n "$differences" ]; then
        echo "Differences found in output$i.txt and ctest$i.txt:"
        echo "$differences"
    else
        echo "input$i.txt: OK"
    fi
done


Overwriting torture_find_diference.sh


In [600]:
!chmod u+x torture_find_diference.sh
!./torture_find_diference.sh

-----------------------------------
Running find_cpp with input14.txt
360
input14.txt: OK
-----------------------------------
Running find_cpp with input15.txt
20
input15.txt: OK
-----------------------------------
Running find_cpp with input16.txt
344378000
input16.txt: OK
-----------------------------------
Running find_cpp with input17.txt
0
input17.txt: OK
-----------------------------------
Running find_cpp with input18.txt
8
input18.txt: OK
-----------------------------------
Running find_cpp with input19.txt
90
input19.txt: OK
-----------------------------------
Running find_cpp with input20.txt
20
input20.txt: OK
-----------------------------------
Running find_cpp with input21.txt
1109889
input21.txt: OK
-----------------------------------
Running find_cpp with input22.txt
10989
input22.txt: OK
-----------------------------------
Running find_cpp with input23.txt
342
input23.txt: OK
-----------------------------------
Running find_cpp with input24.txt
72825312
input24.txt: OK


Count executed instructions in code and verify with inputs 1,2 and 3


In [592]:
%%writefile max_pairwise_product_long_sort_instructions.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cmath>
#include <algorithm>


int main() {
    int instrucciones = 0;
    int n;
    instrucciones++;

    std::cin >> n;
    instrucciones++;

    std::vector<int> valores(n);
    instrucciones++;

    for (int i = 0; i < n; ++i) {
        std::cin >> valores[i];
        instrucciones++;
    }

    std::sort(valores.begin(), valores.end());
    instrucciones = instrucciones + (n * log(n));

    std::cout << valores[n-1] * valores[n-2] << "\n";
    std::cout << instrucciones << "\n";
    return 0;
}

Overwriting max_pairwise_product_long_sort_instructions.cpp


In [587]:
%%writefile max_pairwise_product_long_sort_instructions.sh
#!/bin/bash

g++ max_pairwise_product_long_sort_instructions.cpp -o max_pairwise_product_long_sort_instructions
for i in {1..3}
do
    echo "-----------------------------------"
    echo "Count executed instructions with input$i.txt"

    ./max_pairwise_product_long_sort_instructions < input$i.txt

done

Overwriting max_pairwise_product_long_sort_instructions.sh


In [593]:
!chmod u+x max_pairwise_product_long_sort_instructions.sh
!./max_pairwise_product_long_sort_instructions.sh

-----------------------------------
Count executed instructions with input1.txt
90
36
-----------------------------------
Count executed instructions with input2.txt
90
36
-----------------------------------
Count executed instructions with input3.txt
1345294336
19


Calculate estimated time by instruction and number of instructions by second

In [594]:
%%time
!g++ max_pairwise_product_long_sort_instructions.cpp -o max_pairwise_product_long_sort_instructions
!./max_pairwise_product_long_sort_instructions < input1.txt

90
36
CPU times: user 18 ms, sys: 794 µs, total: 18.8 ms
Wall time: 1.11 s


In [ ]:
0.0167 /36

0.0004638888888888889

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [595]:
%%writefile max_pairwise_sort_efficient.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cmath>
#include <algorithm>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;
    cin >> n;                                 // 1 instrucción: Leer un entero desde la entrada estándar.
    int maxValor = INT_MIN;                   // 1 instrucción: Inicializar maxValor.
    int segMaxValor = INT_MIN;                // 1 instrucción: Inicializar segMaxValor.

    vector<int> numbers(n);                   // n instrucciones: Crear un vector e inicializarlo.

    for (int i = 0; i < n; ++i) {             // n instrucciones: Leer y almacenar valores en el vector.
        cin >> numbers[i];
    }

    for (int i = 0; i < n; ++i) {             // n instrucciones: Realizar un bucle for.
        int aux = numbers[i];                 // 1 instrucción: Asignar un valor a la variable aux.

        if (aux > maxValor) {                 // 1 instrucción: Comparar aux con maxValor.
            segMaxValor = maxValor;           // 1 instrucción: Asignar maxValor a segMaxValor.
            maxValor = aux;                   // 1 instrucción: Asignar aux a maxValor.
        } else if (aux > segMaxValor) {       // 1 instrucción: Comparar aux con segMaxValor.
            segMaxValor = aux;                // 1 instrucción: Asignar aux a segMaxValor.
        }
    }

    cout << "Product: " << maxNumber * secondMaxNumber << "\n"; // 2 instrucciones: Realizar la multiplicación y la salida.
    return 0;                                  // 1 instrucción: Devolver un valor.
}

Overwriting max_pairwise_sort_efficient.cpp


In [596]:
%%writefile max_pairwise_sort_efficient.py

def max_pairwise_sort_efficient(valores):
    n = len(valores)                          # 1 instrucción: Obtener la longitud de la lista "valores".
    valores.sort()                            # n * log(n) instrucciones: Ordenar la lista "valores".

    return valores[n - 1] * valores[n - 2]    # 2 instrucciones: Realizar la multiplicación y la resta.

n = int(input())                              # 1 instrucción: Leer un entero desde la entrada estándar.

valores = [int(i) for i in input().split()]  # n instrucciones: Leer y convertir valores desde la entrada estándar.
print(max_pairwise_sort_efficient(valores))  # 1 instrucción: Llamar a la función e imprimir el resultado.

Overwriting max_pairwise_sort_efficient.py


Test with inputs 1 and 2



In [615]:
!python3 max_pairwise_sort_efficient.py < input1.txt
!python3 max_pairwise_sort_efficient.py < input2.txt

90
90


Test and time it with input 3

In [616]:
!time python3 max_pairwise_sort_efficient.py < input3.txt

40000000000

real	0m0.077s
user	0m0.050s
sys	0m0.011s


20 stress tests

In [602]:
 %%writefile torture_find_diference_python.sh
#!/bin/bash

for i in {14..34}
do
    echo "-----------------------------------"
    echo "Running find_cpp with input$i.txt"
    python max_pairwise_sort_efficient.py < input$i.txt>output$i.txt
    python max_pairwise_sort_efficient.py < input$i.txt
    differences=$(grep -xvFf output$i.txt ctest$i.txt)
    if [ -n "$differences" ]; then
        echo "Differences found in output$i.txt and ctest$i.txt:"
        echo "$differences"
    else
        echo "input$i.txt: OK"
    fi
done

Overwriting torture_find_diference_python.sh


In [603]:
!chmod u+x torture_find_diference_python.sh
!./torture_find_diference_python.sh

-----------------------------------
Running find_cpp with input14.txt
360
input14.txt: OK
-----------------------------------
Running find_cpp with input15.txt
20
input15.txt: OK
-----------------------------------
Running find_cpp with input16.txt
344378000
input16.txt: OK
-----------------------------------
Running find_cpp with input17.txt
0
input17.txt: OK
-----------------------------------
Running find_cpp with input18.txt
8
input18.txt: OK
-----------------------------------
Running find_cpp with input19.txt
90
input19.txt: OK
-----------------------------------
Running find_cpp with input20.txt
20
input20.txt: OK
-----------------------------------
Running find_cpp with input21.txt
1109889
input21.txt: OK
-----------------------------------
Running find_cpp with input22.txt
10989
input22.txt: OK
-----------------------------------
Running find_cpp with input23.txt
342
input23.txt: OK
-----------------------------------
Running find_cpp with input24.txt
72825312
input24.txt: OK


Count executed instructions in code and verify with inputs 1,2 and 3


In [614]:
%%writefile max_pairwise_sort_efficient_instructions.py
import math

def max_pairwise_sort_efficient_instructions(valores):
  global instrucciones
  instrucciones += 1
  n = len(valores)

  valores.sort() # nlgn -- n
  instrucciones += n * math.log(n)
  instrucciones += 1
  return valores[n-1] * valores[n-2]

instrucciones = 0
n = int(input())
instrucciones += 1
valores = [int(i) for i in input().split()]
instrucciones += 1

print("valor: ", max_pairwise_sort_efficient_instructions(valores))
print("instrucciones: ", instrucciones)

Overwriting max_pairwise_sort_efficient_instructions.py


In [619]:
!echo "---input 1------"
! python3 max_pairwise_sort_efficient_instructions.py < input1.txt
!echo "---input 2------"
! python3 max_pairwise_sort_efficient_instructions.py < input2.txt
!echo "---input 3------"
! python3 max_pairwise_sort_efficient_instructions.py < input3.txt

---input 1------
valor:  90
instrucciones:  27.02585092994046
---input 2------
valor:  90
instrucciones:  27.02585092994046
---input 3------
valor:  40000000000
instrucciones:  14.75055681536833


Calculate estimated time by instruction and number of instructions by second

In [620]:
1!time python3 max_pairwise_sort_efficient_instructions.py < input1.txt

valor:  90
instrucciones:  27.02585092994046

real	0m0.046s
user	0m0.035s
sys	0m0.007s


In [573]:
0.045/27

0.0016666666666666666

### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [621]:
%%writefile max_pairwise_sort_super_efficient.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cmath>
#include <algorithm>
#include <climits>

using namespace std;

int main() {

    int n;                                                      // 1 instrucción: Declarar una variable.
    cin >> n;                                                   // 1 instrucción: Leer un entero desde la entrada estándar.
    int maxValor = INT_MIN;                                     // 1 instrucción: Inicializar maxValor.
    int segMaxValor = INT_MIN;                                  // 1 instrucción: Inicializar segMaxValor.

    for (int i = 0; i < n; ++i) {                               // 1 + 2n instrucciones: Realizar un bucle for.
        int aux = 0;                                            // n instrucciones: Declarar y asignar un valor a la variable aux.
        cin >> aux;                                             // n instrucciones: Leer un entero desde la entrada estándar.

        if (aux > maxValor) {                                   // n instrucciones: Comparar aux con maxValor.
            segMaxValor = maxValor;                             // n instrucciones(peor caso): Asignar maxValor a segMaxValor.
            maxValor = aux;                                     // n instrucciones(peor caso): Asignar aux a maxValor.
        } else if (aux > segMaxValor) {                         // n instrucciones: Comparar aux con segMaxValor.
            segMaxValor = aux;                                  // n instrucciones (peor caso): Asignar aux a segMaxValor.
        }
    }

    cout << "Product: " << maxValor * segMaxValor << "\n"; // 1 instrucción: Realizar la multiplicación y la salida.
    return 0;                                                   // 1 instrucción: Devolver un valor.
}

Overwriting max_pairwise_sort_super_efficient.cpp


Test with inputs 1 and 2




In [622]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input1.txt
!./max_pairwise_sort_super_efficient < input2.txt

Product: 90
Product: 90


Test and time it with input 3

In [623]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!time ./max_pairwise_sort_super_efficient < input3.txt

Product: 1345294336

real	0m0.003s
user	0m0.003s
sys	0m0.000s


20 stress tests

In [637]:
 %%writefile torture_max_pairwise_sort_super_efficient.sh
#!/bin/bash

!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
for i in {14..34}
do
    echo "-----------------------------------"
    echo "Running find_cpp with input$i.txt"
    ./max_pairwise_sort_super_efficient < input$i.txt

done


Overwriting torture_max_pairwise_sort_super_efficient.sh


In [638]:
!chmod u+x torture_max_pairwise_sort_super_efficient.sh
!./torture_max_pairwise_sort_super_efficient.sh

./torture_max_pairwise_sort_super_efficient.sh: line 3: !g++: command not found
-----------------------------------
Running find_cpp with input14.txt
Product: 360
-----------------------------------
Running find_cpp with input15.txt
Product: 20
-----------------------------------
Running find_cpp with input16.txt
Product: 344378000
-----------------------------------
Running find_cpp with input17.txt
Product: 0
-----------------------------------
Running find_cpp with input18.txt
Product: 8
-----------------------------------
Running find_cpp with input19.txt
Product: 90
-----------------------------------
Running find_cpp with input20.txt
Product: 20
-----------------------------------
Running find_cpp with input21.txt
Product: 1109889
-----------------------------------
Running find_cpp with input22.txt
Product: 10989
-----------------------------------
Running find_cpp with input23.txt
Product: 342
-----------------------------------
Running find_cpp with input24.txt
Product: 728253

Count executed instructions in code and verify with inputs 1,2 and 3

In [642]:
%%writefile max_pairwise_sort_super_efficient_count.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cmath>
#include <algorithm>
#include <climits>

using namespace std;

int main() {

    int instrucciones = 0;
    int n;
    instrucciones ++;
    cin >> n;
    instrucciones ++;
    int maxValor = INT_MIN;
    instrucciones ++;
    int segMaxValor = INT_MIN;
    instrucciones ++;


    instrucciones ++;
    for (int i = 0; i < n; ++i) {
        instrucciones += 2;
        int aux = 0;
        instrucciones ++;

        cin>> aux;
        instrucciones ++;
        instrucciones ++;
        instrucciones ++;

        if (aux > maxValor){
          segMaxValor = maxValor;
          instrucciones ++;
          maxValor = aux;
          instrucciones++;

        } else if (aux > segMaxValor) {
          segMaxValor = aux;
          instrucciones ++;
        }

    }

    cout <<  "instrucciones: " << instrucciones << " n: "<< n << " Valor: " << maxValor * segMaxValor << "\n";
    return 0;
}

Overwriting max_pairwise_sort_super_efficient_count.cpp


In [643]:
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt
!./max_pairwise_sort_super_efficient_count < input2.txt
!./max_pairwise_sort_super_efficient_count < input3.txt

instrucciones: 85 n: 10 Valor: 90
instrucciones: 68 n: 10 Valor: 90
instrucciones: 49 n: 6 Valor: 1345294336


Calculate estimated time by instruction and number of instructions by second

In [644]:
%%time
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt

instrucciones: 85 n: 10 Valor: 90
CPU times: user 13.8 ms, sys: 3.33 ms, total: 17.1 ms
Wall time: 914 ms


In [613]:
0.0167 / 85

0.0001964705882352941

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [645]:
%%writefile max_pairwise_sort_super_efficient_instructions.py

def max_pairwise_sort_super_efficient_instructions(numbers):
    global max_value
    max_value=float("-inf")
    global second_max_value
    second_max_value=float("-inf")
    global instrucciones
    instrucciones=2

    for num in numbers:
        instrucciones=instrucciones+1
        if num > max_value:
            second_max_value = max_value
            max_value = num
            instrucciones=instrucciones+3
        elif num > second_max_value:
            second_max_value = num
            instrucciones=instrucciones+5
    return max_value*second_max_value

n = int(input())
numbers=[int(i) for i in input().split()]
print("Resultado: ",max_pairwise_sort_super_efficient_instructions(numbers))
print("Instrucciones: ",instrucciones+1)


Overwriting max_pairwise_sort_super_efficient_instructions.py


Test with inputs 1 and 2




In [629]:
! python3 max_pairwise_sort_super_efficient_instructions.py < input1.txt
! python3 max_pairwise_sort_super_efficient_instructions.py < input2.txt

Resultado:  90
Instrucciones:  43
Resultado:  90
Instrucciones:  21


Test and time it with input 3

In [630]:
! python3 max_pairwise_sort_super_efficient_instructions.py < input3.txt

Resultado:  40000000000
Instrucciones:  28


20 stress tests

In [647]:
 %%writefile torture_max_pairwise_sort_super_efficient_python.sh
#!/bin/bash

for i in {14..34}
do
    echo "-----------------------------------"
    echo "Running find_cpp with input$i.txt"
    python max_pairwise_sort_super_efficient_instructions.py < input$i.txt

done


Writing torture_max_pairwise_sort_super_efficient_python.sh


In [648]:
!chmod u+x torture_max_pairwise_sort_super_efficient_python.sh
!./torture_max_pairwise_sort_super_efficient_python.sh

-----------------------------------
Running find_cpp with input14.txt
Resultado:  360
Instrucciones:  43
-----------------------------------
Running find_cpp with input15.txt
Resultado:  20
Instrucciones:  23
-----------------------------------
Running find_cpp with input16.txt
Resultado:  344378000
Instrucciones:  63
-----------------------------------
Running find_cpp with input17.txt
Resultado:  0
Instrucciones:  21
-----------------------------------
Running find_cpp with input18.txt
Resultado:  8
Instrucciones:  27
-----------------------------------
Running find_cpp with input19.txt
Resultado:  90
Instrucciones:  43
-----------------------------------
Running find_cpp with input20.txt
Resultado:  20
Instrucciones:  16
-----------------------------------
Running find_cpp with input21.txt
Resultado:  1109889
Instrucciones:  19
-----------------------------------
Running find_cpp with input22.txt
Resultado:  10989
Instrucciones:  15
-----------------------------------
Running find_c

Calculate estimated time by instruction and number of instructions by second

In [649]:
!time python3 max_pairwise_sort_super_efficient_instructions.py < input1.txt

Resultado:  90
Instrucciones:  43

real	0m0.043s
user	0m0.034s
sys	0m0.007s


In [ ]:
0.036/43

0.0008372093023255813